graphs/mis/sync_metivier_et_al/sync_metivier_et_al.go

```diff
@@ -0,0 +1,349 @@
+			neighbor.Wins[neighbor.WinPhase] = !bit // we know whether we won this phase against this neighbor
+			neighbor.WinPhase++                     // we can start the next phase of symmetry breaking
+			neighbor.BitIndex = 0
+		} else { // still symmetrical
+			neighbor.BitIndex++ // we will have to send another bit
+		}
+	}
+}
+
+func (s *state) sendIns(node lib.Node) {
+	in := s.canEnterMIS()
+	for i, neighbor := range s.Neighbors {
+		if neighbor.Responsive {
+			sendMessage(node, i, in)
+		}
+	}
+	if in != nil {
+		s.InsSent = true
+		s.Chosen = *in
+	}
+}
+
+func (s *state) receiveIns(node lib.Node) {
+	for i, neighbor := range s.Neighbors {
+		if !neighbor.Responsive {
+			continue
+		}
+		in := receiveMessage(node, i)
+		if in != nil {
+			neighbor.Ins[neighbor.InPhase] = *in
+			neighbor.InPhase++
+		}
+	}
+}
+
+func (s *state) sendOuts(node lib.Node) {
+	out := s.shouldDisconnect()
+	for i, neighbor := range s.Neighbors {
+		if neighbor.Responsive {
+			sendMessage(node, i, out)
+		}
+	}
+	if out != nil {
+		s.OutsSent = true
+		s.Done = *out
+	}
+}
+
+func (s *state) receiveOuts(node lib.Node) {
+	for i, neighbor := range s.Neighbors {
+		if !neighbor.Responsive {
+			continue
+		}
+		out := receiveMessage(node, i)
+		if out != nil {
+			neighbor.Responsive = !*out
+			neighbor.Outs[neighbor.OutPhase] = *out
+			neighbor.OutPhase++
+		}
+	}
+}
+
+func (s *state) getNeighborBit(neighbor *neighbor) bool {
+	return s.getBit(neighbor.WinPhase, neighbor.BitIndex)
+}
+
+func (s *state) getBit(phase int, index int) bool {
+	bits := s.Bits[phase]
+	if index > len(bits) {
+		panic("Earlier bits have not been drawn yet")
+	}
+	if index == len(bits) {
+		rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+		bits = append(bits, rng.Intn(2) == 1)
+	}
+	s.Bits[phase] = bits
+	return bits[index]
+}
+
+func (s *state) canEnterMIS() *bool {
+	if s.InsSent {
+		return nil
+	}
+	all := true
+	enter := true
+	for _, neighbor := range s.Neighbors {
+		win, ok := neighbor.Wins[s.Phase]
+		if !ok {
+			all = false
+		} else if !win {
+			enter = false
+		}
+	}
+	if !all {
+		return nil
+	}
+	return &enter
+}
+
+func (s *state) shouldDisconnect() *bool {
+	if s.OutsSent {
+		return nil
+	}
+	all := true
+	disconnect := false
+	for _, neighbor := range s.Neighbors {
+		in, ok := neighbor.Ins[s.Phase]
+		if !ok {
+			all = false
+		} else if in {
+			disconnect = true
+		}
+	}
+	if !all {
+		return nil
+	}
+	if s.Chosen {
+		disconnect = true
+	}
+	return &disconnect
+}
+
+func (s *state) advancePhase() {
+	all := true
+	for _, neighbor := range s.Neighbors {
+		_, ok := neighbor.Outs[s.Phase]
+		if !ok {
+			all = false
+		}
+	}
+	if !all {
+		return
+	}
+	for i, neighbor := range s.Neighbors {
+		if neighbor.Outs[s.Phase] {
+			delete(s.Neighbors, i)
+		} else {
+			delete(neighbor.Wins, s.Phase)
+			delete(neighbor.Ins, s.Phase)
+			delete(neighbor.Outs, s.Phase)
+		}
+	}
+	delete(s.Bits, s.Phase)
+	s.Phase++
+	s.InsSent = false
+	s.OutsSent = false
```

Comment:
`s.InsSent, s.OutsSent = false, false` should be OK

Reply:
I always forget that Go allows that. Done